src/cpp/session/modules/SessionCopilot.cpp

```diff
@@ -0,0 +1,894 @@
+         std::cerr << "REQUEST" << std::endl;
+         std::cerr << "----------------" << std::endl;
+         std::cerr << request << std::endl;
+         std::cerr << "----------------" << std::endl;
+         std::cerr << std::endl << std::endl;
+      }
+
+      operations.writeToStdin(request, false);
+   }
+
+   return true;
+}
+
+void onStdout(ProcessOperations& operations, const std::string& stdOut)
+{
+   // Discard empty lines.
+   static const boost::regex reWhitespace("^\\s*$");
+   if (regex_utils::match(stdOut, reWhitespace))
+      return;
+
+   // Copilot responses will have the format
+   //
+   //    Content-Length: xyz
+   //
+   //    <body>
+   //
+   // Importantly, there is no character separating the body from the
+   // next response, so we're forced to parse headers here.
+   //
+   // TODO: Does this need to be a state machine? Is it possible that we
+   // receive some incomplete requests here?
+   std::size_t startIndex = 0;
+
+   while (startIndex < stdOut.length())
+   {
+      // Find the split.
+      auto splitIndex = stdOut.find("\r\n\r\n", startIndex);
+      if (splitIndex == std::string::npos)
+      {
+         ELOG("Internal error: parsing response failed.");
+         break;
+      }
+
+      // Extract the header text.
+      std::string headerText = string_utils::substring(stdOut, startIndex, splitIndex + 4);
+
+      // Parse the headers.
+      core::http::Headers headers;
+      std::istringstream iss(headerText);
+      core::http::parseHeaders(iss, &headers);
+
+      // Check for content length.
+      std::string contentLength = core::http::headerValue(headers, "Content-Length");
+      if (contentLength.empty())
+      {
+         ELOG("Internal error: response contains no Content-Length header.");
+         break;
+      }
+
+      // Consume that text.
+      auto bodyStart = splitIndex + 4;
+      auto bodyEnd = bodyStart + safe_convert::stringTo<int>(contentLength, 0);
+      std::string bodyText = string_utils::substring(stdOut, bodyStart, bodyEnd);
+      s_pendingResponses.push(bodyText);
+
+      if (copilotLogLevel() > 2)
+      {
+         std::cerr << std::endl;
+         std::cerr << "RESPONSE" << std::endl;
+         std::cerr << "------------------" << std::endl;
+         std::cerr << bodyText << std::endl;
+         std::cerr << "------------------" << std::endl;
+         std::cerr << std::endl << std::endl;
+      }
+
+      // Update the start index.
+      startIndex = bodyEnd;
+   }
+}
+
+void onStderr(ProcessOperations& operations, const std::string& stdErr)
+{
+   std::cerr << stdErr << std::endl;
+}
+
+void onExit(int status)
+{
+   s_agentPid = -1;
+
+   if (status != 0)
+   {
+      ELOG("Agent exited with status {}.", status);
+   }
+}
+
+} // end namespace agent
+
+void stopAgent()
+{
+   if (s_agentPid == -1)
+   {
+      DLOG("No agent running; nothing to do.");
+      return;
+   }
+
+   Error error = core::system::terminateProcess(s_agentPid);
```

Comment:
Should we note here that the onExit callback is responsible for unsetting the `s_agentPid` value here? Or should we do that here instead?